### PR TITLE
[SPARK-41713][SQL] Make CTAS hold a nested execution for data writing

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

This pr aims to make ctas use a nested execution instead of running data writing cmmand.

So, we can clean up ctas itself to remove the unnecessary v1write information. Now, the v1writes only have two implementation: `InsertIntoHadoopFsRelationCommand` and `InsertIntoHiveTable`

### Why are the changes needed?

Make v1writes code clear.

```sql
EXPLAIN FORMATTED CREATE TABLE t2 USING PARQUET AS SELECT * FROM t;

== Physical Plan ==
Execute CreateDataSourceTableAsSelectCommand (1)
   +- CreateDataSourceTableAsSelectCommand (2)
         +- Project (5)
            +- SubqueryAlias (4)
               +- LogicalRelation (3)

(1) Execute CreateDataSourceTableAsSelectCommand
Output: []

(2) CreateDataSourceTableAsSelectCommand
Arguments: `spark_catalog`.`default`.`t2`, ErrorIfExists, [c1, c2]

(3) LogicalRelation
Arguments: parquet, [c1#11, c2#12], `spark_catalog`.`default`.`t`, org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe, false

(4) SubqueryAlias
Arguments: spark_catalog.default.t

(5) Project
Arguments: [c1#11, c2#12]
```

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

improve existed test

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala [hunk 1]
```diff
@@ -143,27 +141,7 @@ case class CreateDataSourceTableAsSelectCommand(
...
-      outputColumns,
-      query,
-      SparkSession.active.sessionState.conf.resolver)
-  }
-
-  override def requiredOrdering: Seq[SortOrder] = {
-    val options = table.storage.properties
-    V1WritesUtils.getSortOrder(outputColumns, partitionColumns, table.bucketSpec, options)
-  }
+  extends DataWritingCommand {
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala:144]: we can make ctas inherit `LeafRunnableCommand` after we move metrics to `WriteFilesExec`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala [hunk 2]
```diff
@@ -143,29 +141,9 @@ case class CreateDataSourceTableAsSelectCommand(
...
-      outputColumns,
-      query,
-      SparkSession.active.sessionState.conf.resolver)
-  }
-
-  override def requiredOrdering: Seq[SortOrder] = {
-    val options = table.storage.properties
-    V1WritesUtils.getSortOrder(outputColumns, partitionColumns, table.bucketSpec, options)
-  }
+  extends LeafRunnableCommand {
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala:144]: this is the key change. now ctas is not a v1 write command.

**Comment by ulysses-you** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala [hunk 3]
```diff
@@ -143,29 +141,9 @@ case class CreateDataSourceTableAsSelectCommand(
     mode: SaveMode,
     query: LogicalPlan,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala:142]: how do we guarantee this is resolved?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala [hunk 4]
```diff
@@ -21,43 +21,23 @@ import scala.util.control.NonFatal
...
-import org.apache.spark.sql.execution.SparkPlan
-import org.apache.spark.sql.execution.command.{DataWritingCommand, DDLUtils}
-import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, InsertIntoHadoopFsRelationCommand, LogicalRelation, V1WriteCommand, V1WritesUtils}
+import org.apache.spark.sql.execution.command.{DataWritingCommand, DDLUtils, LeafRunnableCommand}
+import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, InsertIntoHadoopFsRelationCommand, LogicalRelation}
 import org.apache.spark.sql.hive.HiveSessionCatalog
 import org.apache.spark.util.Utils
 
-trait CreateHiveTableAsSelectBase extends V1WriteCommand with V1WritesHiveUtils {
+trait CreateHiveTableAsSelectBase extends LeafRunnableCommand {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala:32]: do we still need `OptimizedCreateHiveTableAsSelectCommand`? The nested `InsertIntoHadoopFsRelationCommand` should be optimized instead.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala (hunk 3 above)
Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala:142]: It must be resolved before going to `CreateDataSourceTableAsSelectCommand`.

https://github.com/apache/spark/blob/030c1baf2bd13f392feceeb02fc6fed7a4c07b93/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala#L142-L144

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala (hunk 4 above)
Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala:32]: good question, but if we do not have a `OptimizedCreateHiveTableAsSelectCommand`, how can we get `InsertIntoHadoopFsRelationCommand` ..

The pipeline is:  `CreateHiveTableAsSelectCommand` -> `OptimizedCreateHiveTableAsSelectCommand` -> `InsertIntoHadoopFsRelationCommand`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala:142]: can we add an assert then?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala:32]: The new pipeline can be `CreateHiveTableAsSelectCommand` -> `hive insertion command` -> `InsertIntoHadoopFsRelationCommand`

Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala:32]: how to deal with config `spark.sql.hive.convertMetastoreCtas` ? we do not know if the hive insertion is from ctas.

Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala:32]: or, just deprecated this config

Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala:32]: remove `OptimizedCreateHiveTableAsSelectCommand` is out of the scope of this pr, how about doing this in a separated pr ?

Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala:32]: sure

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 5]
```diff
@@ -1217,7 +1230,7 @@ class AdaptiveQueryExecSuite
                 assert(planInfo.nodeName == "AdaptiveSparkPlan")
                 assert(planInfo.children.size == 1)
                 assert(planInfo.children.head.nodeName ==
-                  "Execute CreateDataSourceTableAsSelectCommand")
+                  "Execute InsertIntoHadoopFsRelationCommand")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1233]: shall we check 2 items? One is `CreateDataSourceTableAsSelectCommand` and the other is `InsertIntoHadoopFsRelationCommand`

File: sql/core/src/test/scala/org/apache/spark/sql/util/DataFrameCallbackSuite.scala [hunk 6]
```diff
@@ -217,10 +217,10 @@ class DataFrameCallbackSuite extends QueryTest
     withTable("tab") {
       spark.range(10).select($"id", $"id" % 5 as "p").write.partitionBy("p").saveAsTable("tab")
       sparkContext.listenerBus.waitUntilEmpty()
-      assert(commands.length == 5)
-      assert(commands(4)._1 == "command")
-      assert(commands(4)._2.isInstanceOf[CreateDataSourceTableAsSelectCommand])
-      assert(commands(4)._2.asInstanceOf[CreateDataSourceTableAsSelectCommand]
+      assert(commands.length == 6)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/util/DataFrameCallbackSuite.scala:220]: shall we add a comment to explain it?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveExplainSuite.scala [hunk 7]
```diff
@@ -102,21 +102,15 @@ class HiveExplainSuite extends QueryTest with SQLTestUtils with TestHiveSingleto
 
   test("explain create table command") {
     checkKeywordsExist(sql("explain create table temp__b using hive as select * from src limit 2"),
-                   "== Physical Plan ==",
-                   "InsertIntoHiveTable",
-                   "Limit",
-                   "src")
+                   "== Physical Plan ==")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveExplainSuite.scala:105]: shall we at least check something?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala [hunk 8]
```diff
@@ -143,29 +141,10 @@ case class CreateDataSourceTableAsSelectCommand(
...
-      query,
-      SparkSession.active.sessionState.conf.resolver)
-  }
-
-  override def requiredOrdering: Seq[SortOrder] = {
-    val options = table.storage.properties
-    V1WritesUtils.getSortOrder(outputColumns, partitionColumns, table.bucketSpec, options)
-  }
+  extends LeafRunnableCommand {
+  assert(query.resolved)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala:146]: shall we mark the input query as `innerChildren`? Then people can still see it via EXPLAIN. The same to hive CTAS.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala [hunk 9]
```diff
@@ -143,29 +141,10 @@ case class CreateDataSourceTableAsSelectCommand(
     mode: SaveMode,
     query: LogicalPlan,
     outputColumnNames: Seq[String])
-  extends V1WriteCommand {
-
-  override def fileFormatProvider: Boolean = {
-    table.provider.forall { provider =>
-      classOf[FileFormat].isAssignableFrom(DataSource.providingClass(provider, conf))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala:150]: shall we revert the change in `DataSource.scala` as well?

**Comment by ulysses-you** [comment]
@cloud-fan addressed all comments

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala [hunk 10]
```diff
@@ -143,29 +141,11 @@ case class CreateDataSourceTableAsSelectCommand(
...
-      SparkSession.active.sessionState.conf.resolver)
-  }
-
-  override def requiredOrdering: Seq[SortOrder] = {
-    val options = table.storage.properties
-    V1WritesUtils.getSortOrder(outputColumns, partitionColumns, table.bucketSpec, options)
-  }
+  extends LeafRunnableCommand {
+  assert(query.resolved)
+  override def innerChildren: Seq[LogicalPlan] = query :: Nil
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala:146]: can we put the EXPLAIN  result of a CTAS in the PR description as an example?

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala:146]: sure, updated

**Comment by cloud-fan** [comment]
thanks, merging to master!
